iceoryx_hoofs/include/iceoryx_hoofs/cxx/functional_interface.hpp

```diff
@@ -0,0 +1,279 @@
+{
+};
+
+template <typename Derived>
+struct Expect
+{
+    /// @brief Expects that the object is valid, otherwise the method prints the
+    ///        provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    void expect(const char* const msg) const noexcept;
+};
+
+template <typename Derived, typename ValueType>
+struct ExpectWithValue
+{
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    /// @return a reference to the contained value
+    ValueType& expect(const char* const msg) & noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    /// @return a const reference the contained value
+    const ValueType& expect(const char* const msg) const& noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    /// @return rvalue reference to the contained value
+    ValueType&& expect(const char* const msg) && noexcept;
+
+    /// @brief Expects that the object is valid and returns the contained value, otherwise
+    //         the method prints the provided message and induces a fatal error
+    /// @param[in] msg Message which will be printed when the object is invalid
+    /// @return const rvalue reference to the contained value
+    const ValueType&& expect(const char* const msg) const&& noexcept;
+};
+
+template <typename Derived, typename ValueType>
+struct ValueOr
+{
+    /// @brief When the object contains a value a copy will be returned otherwise a
+    ///        a copy of alternative will be returned
+    /// @param[in]  alternative the return value which will be used when the object does
+    ///             not contain a value
+    /// @return A copy of the contained value if possible otherwise a copy of alternative
+    ValueType value_or(const ValueType& alternative) const noexcept;
+};
+
+template <typename Derived, typename ValueType>
+struct AndThenWithValue
+{
+    using and_then_callback_t = cxx::function_ref<void(ValueType&)>;
+    using const_and_then_callback_t = cxx::function_ref<void(const ValueType&)>;
+
+    /// @brief Calls the provided callable when the object is valid and provides the underlying
+    ///        value reference as argument to the callable. If the object is not valid, nothing
+    ///        happens.
+    /// @param[in] callable Callable to be called when valid
```

Comment:
Not quite sure how this will be rendered and I think we do not follow a common style here but is looks a bit weird to repeat the `callable` as `Callable`.

Don't care that much which style to use but it should be consistent. Do you know it clang tidy could check this for us or do we (finger on my nose) one day need to check the whole documentation for this?

Reply:
I rephrased it into less weird